File: vtk-rs/libvtkrs/src/vtk_vector_text.cpp
#include "vtk_vector_text.h"
#include "vtk_vector_text.rs.h"

vtkVectorText* vtk_vector_text_new() {
    return vtkVectorText::New();
}

void vtk_vector_text_delete(vtkVectorText& ptr) {
    ptr.Delete();
}

void vector_text_set_text(vtkVectorText& source, rust::String text) {
    std::string cpp_text(text.begin(), text.end());
    source.SetText(cpp_text.c_str());
}

rust::String vector_text_get_text(vtkVectorText& source) {
    const char* text = source.GetText();
    return rust::String(text ? text : "");
}

vtkAlgorithmOutput* vector_text_get_output_port(vtkVectorText& source) {
    return source.GetOutputPort();
}

File: vtk-rs/src/vtk_vector_text.rs
use std::pin::Pin;

use crate::algorithm_output_port::AlgorithmOutputPort;

#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("vtk_vector_text.h");

        type vtkVectorText;

        fn vtk_vector_text_new() -> *mut vtkVectorText;
        fn vtk_vector_text_delete(ptr: Pin<&mut vtkVectorText>);

        fn vector_text_set_text(source: Pin<&mut vtkVectorText>, text: String);
        fn vector_text_get_text(source: Pin<&mut vtkVectorText>) -> String;

        fn vector_text_get_output_port(source: Pin<&mut vtkVectorText>) -> *mut vtkAlgorithmOutput;

        type vtkAlgorithmOutput;
    }
}

/// VectorText creates 3D polygonal text that can be used with Follower for billboard labels.
/// Unlike TextSource (which creates extruded text), VectorText creates stroke-based text
/// that's perfect for labels and annotations.
pub struct VectorText {
    ptr: *mut ffi::vtkVectorText,
}

impl VectorText {
    pub fn new() -> Self {
        let ptr = ffi::vtk_vector_text_new();
        Self { ptr }
    }

    fn as_mut(&mut self) -> Pin<&mut ffi::vtkVectorText> {
        unsafe { Pin::new_unchecked(&mut *self.ptr) }
    }

    pub fn set_text(&mut self, text: &str) {
        ffi::vector_text_set_text(self.as_mut(), text.to_string());
    }

    pub fn get_text(&mut self) -> String {
        ffi::vector_text_get_text(self.as_mut())
    }

    pub fn output_port(&mut self) -> AlgorithmOutputPort {
        let ptr = ffi::vector_text_get_output_port(self.as_mut());
        unsafe { crate::AlgorithmOutputPort::from_raw(ptr as *mut std::ffi::c_void) }
    }
}

impl Drop for VectorText {
    fn drop(&mut self) {
        ffi::vtk_vector_text_delete(self.as_mut());
    }
}
